File: 101-109_small_problems/easy_8/sum_of_sums.rb
# Sum of Sums
# Write a method that takes an Array of numbers and then returns the sum of the sums of each leading subsequence for that Array. You may assume that the Array always contains at least one number.

# Input: An array of numbers
# Output: An integer
# Mental model: Given an array of numbers, calculate the sum

# Examples:

# sum_of_sums([3, 5, 2]) == (3) + (3 + 5) + (3 + 5 + 2) # -> (21)
# sum_of_sums([1, 5, 7, 3]) == (1) + (1 + 5) + (1 + 5 + 7) + (1 + 5 + 7 + 3) # -> (36)
# sum_of_sums([4]) == 4
# sum_of_sums([1, 2, 3, 4, 5]) == 35

# Data structure

# Input is an array. Output is an integer.

# Algorithm
# Create counter
# Create sum variable
# Loop through array
#   Multiply array.length - counter * current_element
#   Add result to sum
#   Add 1 to counter

# Return sum

# Code
def sum_of_sums(arr)
  sum = 0
  counter = 0

  loop do
    sum += (arr.length - counter) * arr[counter]
    counter += 1
    break if arr.length == counter
  end

  sum
end

p sum_of_sums([3, 5, 2]) == (3) + (3 + 5) + (3 + 5 + 2) # -> (21)
p sum_of_sums([1, 5, 7, 3]) == (1) + (1 + 5) + (1 + 5 + 7) + (1 + 5 + 7 + 3) # -> (36)
p sum_of_sums([4]) == 4
p sum_of_sums([1, 2, 3, 4, 5]) == 35

File: exercises/js210_small_problems/list_processing/sumOfSums.js
// Sum of Sums
// Write a function that takes an array of numbers, and returns the sum of the sums of each leading subsequence for that array. You may assume that the array always contains at least one number.

// Input: An array of numbers
// Output: A single number
//
// Rules:
// - Return the sum of sums

// Maybe we can do this with two calls of reduce?
// Keep track of current sum value
// Then we have to get slices of the array - from element to all elements
//
// Algorithm
//
// 1. Reduce array:
//   For each element, we get the index of the current element
//   Add accumulator value to the result of another reduce call:
//     Add inner accumulator + slice of array from 0 to current index + 1

function sumOfSums(arr) {
  return arr.reduce(function(sum, current, index) {
    return sum + arr.slice(0, index + 1).reduce(function(innerSum, innerElem) {
      return innerSum + innerElem;
    });
  });
}

// Examples:

console.log(sumOfSums([3, 5, 2]));        // (3) + (3 + 5) + (3 + 5 + 2) --> 21
console.log(sumOfSums([1, 5, 7, 3]));     // (1) + (1 + 5) + (1 + 5 + 7) + (1 + 5 + 7 + 3) --> 36
console.log(sumOfSums([4]));              // 4
console.log(sumOfSums([1, 2, 3, 4, 5]));  // 35
